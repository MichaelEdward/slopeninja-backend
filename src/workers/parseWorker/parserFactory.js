import cheerio from 'cheerio';
import { AllHtmlEntities } from 'html-entities';

const htmlEntities = new AllHtmlEntities();

export const removeBackSlashes = (text) => {
  const decoded = text ? text.toString().replace(/\\/g, '') : null;
  return decoded;
}

export const xyz = (text) => {
  const decoded = text ? text.toString().replace(/\u0026amp;/gu, '__JAKE__') : null;
  return decoded;
}

export const decodeEntities = (text) => {
  // best way I found to normalize unicode in input
  const unicodeInput = JSON.stringify(JSON.parse(text.toString()));
  return htmlEntities.decode(unicodeInput);
}

export const createHtmlParser = (key, parser, preParser = t => t) => async (html) => {
  const $ = cheerio.load(preParser(html), { decodeEntities: true });

  const keyData = await parser($);

  return {
    [key]: keyData
  };
}

export const createJSONParser = (key, parser, preParser = t => t) => async (data = "{}") => {
  let json = JSON.parse(preParser(data));

  let d;
  if (json) {
    d = json;
  }

  const keyData = await parser(d);

  return {
    [key]: keyData
  };
}
