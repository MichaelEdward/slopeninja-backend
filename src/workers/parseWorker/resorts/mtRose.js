import cheerio from 'cheerio';

import {
  degreeOrNull,
  inchOrNull,
  numberOrNull,
  weatherStatusOrNull,
  liftTrailStatusOrNull,
  notEmptyStringOrNull,
  trailLevelOrNull,
  resortStatusOrNull,
} from '../weatherUtil';

const initialSnow = {
  status: null,
  weatherIcon: null,
  temprature: null,
  baseCondition: null,
  newSnow: null,
  snowDepthBase: null,
  snowDepthSummit: null,
};

const initialLifts = {
  total: null,
  open: null,
};

const initialTrails = {
  total: null,
  open: null,
};

export const parseMtRoseSnow = async ($) => {
  const weatherIcon = $('.sr-current-cond .sr-cc-wrapper .sr-text').first().text().trim();
  const status = $('.sr-mountain-notes.sr-row .sr-text p').first().text().trim();
  const temprature = $('.sr-current-cond.sr-row.row .col-sm-6 .sr-cc-wrapper.row .sr-temp h1').text().trim();
  //24 Hours
  const newSnow24Hr = $('.sr-snow-totals.sr-row.row .sr-snow-total h2').first().text().trim();
  //Base
  // const snowDepthBase = $('.conditions-overlay .row.weather-row .large-4.columns .weather-data').slice(1,2).text().trim();
  // const snowDepthSummit = $('.snowConditions tbody tr td').slice(3,4).text().trim();
  return {
    ...initialSnow,
    weatherIcon: weatherStatusOrNull(weatherIcon),
    status: resortStatusOrNull(status),
    temprature: degreeOrNull(temprature),
    newSnow: numberOrNull(Number.parseInt(newSnow24Hr)),
    // snowDepthBase: inchOrNull(snowDepthBase),
    // snowDepthSummit: inchOrNull(snowDepthSummit),
  };
}

export const parseMtRoseLifts = async ($) => {
  return {
    ...initialLifts,
  };
}

export const parseMtRoseTrails = async ($) => {
  return {
    ...initialTrails,
  };
}

export const parseMtRoseLiftList = async ($) => {
  const list = [];

  $('.sr-lifts-wrapper .sr-ski-lift-wrapper').map((index, rowElement) => {
    const nameText = $(rowElement).find('.sr-lift-name').text().trim();
    const statusText = $(rowElement).find('.sr-lift-status').text().trim();
    //
    const name = notEmptyStringOrNull(nameText);
    const status = liftTrailStatusOrNull(statusText);
    const category = null;

    const lift = {
      name,
      status,
      category,
    };

    list.push(lift)


  });
 return list;
}

export const parseMtRoseTrailList = async ($) => {
  const list = [];

 return list;
}
