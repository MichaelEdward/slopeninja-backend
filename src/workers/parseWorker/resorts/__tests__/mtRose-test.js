import fs from 'fs';
import { parseMtRoseWeather, parseMtRoseLifts, parseMtRoseTrails } from '../mtRose';
import { createHtmlParser } from '../../parserFactory';

test('fetches MtRose weather data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/mtRose-weather.html`);

  const resortData = await createHtmlParser('weather', parseMtRoseWeather)(htmlText);
  expect(resortData).toEqual({
    weather: {
      status: 'closed',
      weatherIcon: 'sunny',
      temprature: 53,
      baseCondition: null,
      newSnow: 0,
      snowDepthBase: null,
      snowDepthSummit: null,
    },
  });
})

test('fetches all null for nonexisting weather values', async () => {
  const resortData = await createHtmlParser('weather', parseMtRoseWeather)('<html></html>');
  expect(resortData).toEqual({
    weather: {
      status: null,
      weatherIcon: null,
      temprature: null,
      baseCondition: null,
      newSnow: null,
      snowDepthBase: null,
      snowDepthSummit: null,
    }
  });
});

test('fetches MtRose lifts data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/mtRose-weather.html`);
  const resortData = await createHtmlParser('lifts', parseMtRoseLifts)(htmlText);
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: null,
    }
  });
});

test('fetches all null for nonexisting lift values', async () => {
  const resortData = await createHtmlParser('lifts', parseMtRoseLifts)('<html></html>');
  expect(resortData).toEqual({
    lifts: {
      total: null,
      open: null,
    }
  });
});

test('fetches MtRose trails data correctly', async () => {
  const htmlText = fs.readFileSync(`${__dirname}/fixtures/mtRose-weather.html`);
  const resortData = await createHtmlParser('trails', parseMtRoseTrails)(htmlText);
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});

test('fetches all null for nonexisting trails values', async () => {
  const resortData = await createHtmlParser('trails', parseMtRoseTrails)('<html></html>');
  expect(resortData).toEqual({
    trails: {
      total: null,
      open: null,
    }
  });
});
