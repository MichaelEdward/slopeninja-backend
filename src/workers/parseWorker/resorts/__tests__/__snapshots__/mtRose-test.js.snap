// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches MtRose lift list correctly 1`] = `
Object {
  "liftList": Array [
    Object {
      "category": null,
      "name": "Northwest 6",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Blazing Zephyr 6*",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Lakeview",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Wizard",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Galena",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Flying Jenny",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Chuter",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Northwest 6",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Blazing Zephyr 6*",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Lakeview",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Wizard",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Galena",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Flying Jenny",
      "status": "closed",
    },
    Object {
      "category": null,
      "name": "Chuter",
      "status": "closed",
    },
  ],
}
`;
